draft-ietf-jsonpath-base.md

```diff
@@ -1317,10 +1327,25 @@ function-expression     = function-name "(" S [function-argument
 function-argument       = comparable / filter-path
 ~~~
 
-Syntactically, a function-expression can occur anywhere where a
-singular-path can occur (exist-expr, comparable).
-A function-expression that employs a function extension that returns a
-nodelist MUST return a singular node to be used in a comparable.
+Any function expressions in a query must be syntactically valid,
+otherwise the JSONPath implementation MUST raise an error
+(see Section {{<synsem-overview}}).
+
+A function expression is syntactically valid if:
+
+* It conforms to the ABNF syntax above.
+* It occurs as a filter-path in an existence test and the function
+is defined to return either a nodelist or a Singular Nodelist.
+* It occurs as a singular-path on either side of a comparison and the function
+is defined to return either a value or a Singular Nodelist.
```

Comment:
So are we not accepting functions which _may_ return a single value but _could_ return multiple values?

This could only be a runtime determination, so static syntactical analysis couldn't find it.  (I realize that's an argument against such functions; just clarifying.)

Reply:
Static checks reduce the number of runtime errors we have to cope with in the spec, so I'd prefer to rule out that kind of function signature. However, if we allowed a function to return a list of values (which we currently don't), then such a function could return a list of size one.